### ISSUE TITLE
Create a fck-nat Terraform module

### ISSUE THREAD

**Issue by AndrewGuenther** [AndrewGuenther/fck-nat#4]

**Comment by nitrocode** [comment]
Instead of creating a new module, it would probably be better to reuse an existing one that supports overriding the ami.

See https://github.com/cloudposse/terraform-aws-dynamic-subnets#input_nat_instance_ami_id

And a complete example

https://github.com/cloudposse/terraform-aws-dynamic-subnets/blob/master/examples/complete/main.tf

**Comment by AndrewGuenther** [comment]
The intent is that the fck-nat AMI will be compatible with any other NAT instance provider, but certain features like high-availability mode require configuration that many modules don't provide. For example, https://github.com/cloudposse/terraform-aws-dynamic-subnets#input_nat_instance_ami_id only runs a single instance rather than an autoscaling group. A fck-nat Terraform module would be a bit more tailor-made to support features like high-availability mode and other features I have in mind for later on, but the AMI should always be compatible out of the box with modules like the one you linked.

**Comment by JulianCBC** [comment]
@AndrewGuenther I've been trying to find a solution for NAT on a VPC that's implementable using Terraform and I found this module: https://github.com/int128/terraform-aws-nat-instance which is essentially `fck-nat` without the AMI or RPMs. (The scripting is _spookily_ similar.)

I'm hacking together some changes to their scripting to resolve a couple of issues I've found in my particular environment that would apply to the HA use-case for `fck-nat`:
* Dropping the EIP requirement if the EC2 instance doing NAT is on a public subnet
* Better handling of the floating ENI
Would you like me to port them over to `fck-nat`?

Following on from that, would you mind if I propose a patch to that module to make it use `fck-nat`'s scripting? The rough plan would be to either use the `fck-nat` AMI or to submodule `fck-nat` into that repository so it can use `fck-nat`'s scripts. This could then make that module the "official" Terraform module for `fck-nat`.

**Comment by AndrewGuenther** [comment]
I'm always happy to accept improvements. And you're welcome to contribute anything from `fck-nat` back over to `terraform-aws-nat-instance`. That said, I don't know if long-term `terraform-aws-nat-instance` will be a good fit as an "official" module for `fck-nat`. The main reason being that I'm hoping to add additional configuration options in the future and make additional improvements which _may_ extend to infrastructure. However, the `fck-nat` AMI should always be able to run in a mode which can support setups like `terraform-aws-nat-instance` has (patching rules being a good example). I think it would definitely simplify that package if it just referred to the `fck-nat` AMI directly, considering from what I can tell it effectively builds a similar host using user data scripts.

**Comment by JulianCBC** [comment]
Yeah, `fck-nat` and `terraform-aws-nat-instance` are functionally identical, the only real differences being:
* some very minor config changes (the module disables ICMP redirects)
* your AMI attaches the ENI for HA mode in the service instead of in the startup script
* your AMI supports a non-HA mode

IMHO both solutions have their places, your all-in-one, deploy-and-go solution is a great way to get started if you're not using Terraform, whereas the Terraform module uses the latest AMI so it'll get updates faster, which could cause breakage if things change too fast.

Anyway, the first round of fixes have been ported over:
* #11 fixes a couple of significant issues that I found while debugging the Terraform module: disabling reverse path protection in the instance and source/destination checks on the ENIs in VPC.
* #12 switches the upstream interface in HA mode to `eth0` instead of the attached interface, removing the requirement to attach an EIP to that interface.

Both are untested in `fck-nat`, so please review carefully, but functionally identical changes to the module work for me in my environment.

**Comment by souravjamwal77** [comment]
+1

**Comment by ahodges22** [comment]
I've thrown together a quick Terraform module to spin up a fck-nat instance. 

https://registry.terraform.io/modules/ahodges22/fck-nat

**Comment by souravjamwal77** [comment]
As per @ahodges22, his TF module should not be used in production. So, does this apply to fck-nat also?

**Comment by AndrewGuenther** [comment]
Whether you feel fck-nat is viable for your production use is up to you. Fck-nat offers a high availability mode which _minimizes_ downtime in the event of issues which may or may not be acceptable for your use case. Give the docs a read and make the decision that is the best for you.

**Comment by ssadler** [comment]
This was super easy in the end, in my case at least:

```terraform
resource "aws_network_interface" "fck-nat-if" {                                       
  subnet_id       = aws_subnet.subnet_public.id                                       
  security_groups = [aws_default_security_group.default_security_group.id]            
                                                                                      
  source_dest_check = false                                                           
}                                                                                     
                                                                                      
                                                                                      
resource "aws_instance" "fck-nat" {                                                   
  ami           = "ami-0d65d4e8bca87cd3f" # Might want to check this...
  instance_type = "t4g.nano"                                                          
  key_name      = "YOUR_KEY_HERE"                                                   
                                                                                      
  network_interface {                                                                 
    network_interface_id = aws_network_interface.fck-nat-if.id                        
    device_index         = 0                                                          
  }                                                                                   
                                                                                      
  tags = {                                                                            
    Name = "${var.project}-${var.env}-fck-nat"                                        
  }                                                                                   
}                                                                                     
```

Then all I had to do was add this into the private routing table: `network_interface_id = aws_network_interface.fck-nat-if.id`

Et voila, you can delete the nat!

**Comment by jirkavrba** [comment]
I found that you can use `data.aws_ami` for convenience to get the latest version of fck-nat AMI

```hcl
data "aws_ami" "fck_nat" {
  filter {
    name   = "name"
    values = ["fck-nat-amzn2-*"]
  }

  owners      = ["568608671756"]
  most_recent = true
}
```

You can later reference it like this:

```hcl
resource "aws_instance" "nat" {
  // ...
  image_id = data.aws_ami.fck_nat.image_id
  // ...
}
```

**Comment by thecasual** [comment]
+1

**Comment by sjlu** [comment]
> I found that you can use `data.aws_ami` for convenience to get the latest version of fck-nat AMI
> 
> ```terraform
> data "aws_ami" "fck_nat" {
>   filter {
>     name   = "name"
>     values = ["fck-nat-amzn2-*"]
>   }
> 
>   owners      = ["568608671756"]
>   most_recent = true
> }
> ```
> 
> You can later reference it like this:
> 
> ```terraform
> resource "aws_instance" "nat" {
>   // ...
>   image_id = data.aws_ami.fck_nat.image_id
>   // ...
> }
> ```

```
  filter {
    name   = "architecture"
    values = ["arm64"]
  }
```

To get arm64!

**Comment by ahmgithubahm** [comment]
+1 from me

**Comment by Devesh-N** [comment]
+1 for me too

**Comment by mediflux95** [comment]
> > I found that you can use `data.aws_ami` for convenience to get the latest version of fck-nat AMI
> > ```terraform
> > data "aws_ami" "fck_nat" {
> >   filter {
> >     name   = "name"
> >     values = ["fck-nat-amzn2-*"]
> >   }
> > 
> >   owners      = ["568608671756"]
> >   most_recent = true
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can later reference it like this:
> > ```terraform
> > resource "aws_instance" "nat" {
> >   // ...
> >   image_id = data.aws_ami.fck_nat.image_id
> >   // ...
> > }
> > ```
> 
> ```
>   filter {
>     name   = "architecture"
>     values = ["arm64"]
>   }
> ```
> 
> To get arm64!

Hello !
I tried this approach but I don't find any amis in any region.

**Comment by AndrewGuenther** [comment]
@RaJiska's module is the blessed/official Terraform module now and I just merged #55 to point to it from the docs. With that, I am going to call this complete! Appreciate all your work @RaJiska!

Docs available in the `develop` release here: https://fck-nat.dev/develop/deploying/#terraform

Even though the docs are in `develop` you should be able to use terraform-aws-fck-nat right now ahead of the 1.3 release.
